### PR TITLE
Feature/beg 81 initial version

### PR DESCRIPTION
**Pull request by aligent-lturner**
Initial version of Prerender.io module.

- Adds `prerender_io_product` indexer
- Adds config settings (enabled/disabled + token) at default/website/store level
- Generates batch recache requests to prerender.io - batch size set to maximum of 1000 urls at a time.

### CODE REVIEW

File: Model/Api/PrerenderClient.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        if (!$this->prerenderConfigHelper->isRecacheEnabled($storeId)) {
+            return;
+        }
+        $token = $this->prerenderConfigHelper->getToken($storeId);
+        if (empty($token)) {
+            $this->logger->info(__('Prerender token is not set'));
+            return;
+        }
+
+        if (count($urls) > self::MAX_URLS) {
```
Comment (jswift) [inline Model/Api/PrerenderClient.php]: I wouldn't bother with this, the chunk logic will still work even if there's only 1 url

File: Model/Api/PrerenderClient.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    private function sendRequest(array $urls, string $token): void
+    {
+        $payload = $this->jsonSerializer->serialize(
+            [
+                'prerenderToken' => $token,
+                'urls' => $urls
+            ]
+        );
+        $this->client->addHeader('content-type', 'application/json');
```
Comment (jswift) [inline Model/Api/PrerenderClient.php]: Does adding the header to the client add it to all future requests or just the next one? If it's all then we could do this in the constructor

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "aligent/magento2-prerender-io",
+    "description": "Prerender.io integration for Magento 2, providing recaching for product URLs",
+    "require": {
```
Comment (jswift) [inline composer.json:4]: should we add the magento coding standard as a dev requirement?

File: Helper/Config.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Store\Model\ScopeInterface;
+
+class Config
+{
+    private const XML_PATH_RECACHE_ENABLED = 'system/prerender_io/enabled';
+    private const XML_PATH_PRERENDER_TOKEN = 'system/prerender_io/token';
+
+    /** @var ScopeConfigInterface  */
```
Comment (jswift) [inline Helper/Config.php:17]: Is there a reason to have these docblocks I'm not aware of?

Comment (quangdo-aligent) [reply Helper/Config.php:17]: Coding standards pipeline requires this.

File: Model/Api/PrerenderClient.php (hunk 1 above)
Comment (aligent-lturner) [reply Model/Api/PrerenderClient.php]: Yeah - good point :+1:

File: Model/Api/PrerenderClient.php (hunk 2 above)
Comment (aligent-lturner) [reply Model/Api/PrerenderClient.php]: Looks like it's persistent - I'll move it up.

File: Helper/Config.php (hunk 4 above)
Comment (jswift) [reply Helper/Config.php:17]: Ah I can see that now. Kinda assumed they'd drop that rule since it now requires PHP 7 but anyways...

**Review by jswift — APPROVED** [review]

**Pull request merged**